snuba/query/snql/parser.py

```diff
@@ -1350,10 +1351,13 @@ def _post_process(
 
 def parse_snql_query(
     body: str, dataset: Dataset, custom_processing: Optional[CustomProcessors] = None,
-) -> Union[CompositeQuery[QueryEntity], LogicalQuery]:
+) -> Tuple[Union[CompositeQuery[QueryEntity], LogicalQuery], str]:
     with sentry_sdk.start_span(op="parser", description="parse_snql_query_initial"):
         query = parse_snql_query_initial(body)
 
+    with sentry_sdk.start_span(op="parser", description="anonymize_snql_query"):
+        snql_anonymized = format_snql_anonymized(query).get_sql()
+
```

Comment:
Are you returning the anonymized query from the parser in order to format it before applying the post processing steps ?
I am not convinced this is the right approach. We are now coupling the parser with the concept of query log (that needs anonymization). It is very arbitrary that the parser needs to concern itself with how we anonymize the query for logging. That should be a concern internal to the logger. 
In principle, in order to keep your design sane, you want each component, class, function, to depend on as few things as possible, and each dependency is supposed to have a logical meaning. This allows you to understand the design much better and to architect changes. A dependency here on logic that is only concerned with logging is not a dependency that is functional or logically related with parsing.

The issue is that separating these two concerns is going to be tricky. @evanh  do you remember why we do post processing inside the `parse_snql_query` function instead of attaching them to the execution pipeline once the request is built ? Is there a reason or is it there for historical reasons, or was it an arbitrary distinction between what is a "second pass of the parser" vs what is query processing ?

Reply:
I believe it was a mix of both. For example some of the post processing is validation, which we need to do in the parser since subsequent steps don't have a mechanism for rejecting queries. Also the post processing includes steps such as picking the entity for Discover queries and aligning the date conditions, which have an impact on downstream pipeline steps. Some of the post processing was definitely me deciding that it made more sense in the parser instead of adding another query processor.

Conceptually right now we have an idea that at every step of the pipeline, the query is correct and valid. If we split the logic into two steps, the parsing and the post processing, then we will be breaking that contract.

We do want the anonymized query to be directly after parsing to the AST, in order to keep as close as possible to what is actually being sent over the wire. Inserting the logging in here means we don't have to parse the SnQL twice and ensures that what is logged hasn't been transformed by the post processors at all.

Reply:
mh ok, I don't think mixing the two concepts is a great idea, though probably not worth blocking this. 
Can we record a task to revisit these dependencies ?